Welcome chapter (#176)
bevyengine/bevy-website: content/learn/book/welcome/app/_index.md

+++
title = "Creating Bevy apps"
weight = 2
template = "book-section.html"
page_template = "book-section.html"
+++

Bevy programs store and execute all of their game logic and data with a single `App` data structure.

> Let's make a trivial Hello World app.

Reviewer: This seems like it should precede the first line.

I would like to see the flow if "Bevy programs store and execute all of their game logic and data with a single `App` data structure." was moved to below the code block, so the reader know's we're making a trivial game, sees the code, then sees the explanation for said code.

Author: I'd like to go explanation -> example in the book; the quick start guides will be better suited for readers who prefer the opposite direction.

```rust
use bevy::prelude::*;

fn main(){
  App::new()
  .add_system(hello.system())
  .run();
}

fn hello(){
    println!("Hello, Bevy!")
}
```

> The process is straightforward: we first create a new [`App`](https://docs.rs/bevy/latest/bevy/app/struct.App.html).

Reviewer: "Bevy programs store and execute all of their game logic and data with a single `App` data structure."

The quoted text should go after this line, imho.

Then, we add a simple system, which prints "Hello, Bevy!" when it is run.
Finally once we're done configuring the app, we call [`App::run()`](https://docs.rs/bevy/latest/bevy/app/struct.App.html#method.run) to actually make our app *do things*.

## What makes an App?

So, what sort of data does are `App` really store?

> Turns out: it's fairly simple.

Reviewer: Alternatively, or maybe more appropriately, "Bevy programs store and execute all of their game logic and data with a single `App` data structure." should go here?

Looking at the docs for [`App`](https://docs.rs/bevy/latest/bevy/app/struct.App.html), we find three fields: `world`, `runner` and `schedule`.

> You can read more about these by exploring the reference documentation linked just above.

Reviewer: This seems implied by the link present in the previous sentence.

Author: I wanted a bit of a firmer nudge here, targeted towards beginners to programming. "Read the docs" is incredibly powerful but not always intuitive for new programmers or those coming from ecosystems with weak docs.

> Generally, you'll be operating at a higher-level of abstraction than these basic primitives: thinking about data in terms of specific resources or components, adding systems via convenient methods rather than manually supplying a `Schedule`, and sticking to the default runner which runs your game logic in a loop.

Reviewer: higher-level -> higher level

Reviewer: ```suggestion
Generally, you'll be operating at a higher-level of abstraction. You'll likely be thinking about data in terms of specific resources or components, or adding game logic systems via methods to be run in the [default runner](?url?).
```

Also, is "default runner" a Bevy specific term? Are there docs for this to be linked to?
url?? == https://docs.rs/bevy/0.5.0/bevy/app/struct.App.html#structfield.runner

Author: I think the default runner talk is mostly just confusing. I cut it.

> To do so, you'll customize your own `App` using various methods with the builder pattern.

Reviewer: Link to builder pattern? 

https://en.wikipedia.org/wiki/Builder_pattern#:~:text=The%20builder%20pattern%20is%20a,Gang%20of%20Four%20design%20patterns.

Author: I agree that this deserves a link. That wikipedia page is terrible though; let me try and dig up something better.

Author: https://doc.rust-lang.org/1.0.0/style/ownership/builders.html is much better, and Rust specific!

+++
title = "Creating Bevy apps"
weight = 2
template = "book-section.html"
page_template = "book-section.html"
+++

> When you're making a game in Bevy, you will have a single `App`, which stores and handles all of your game logic and data.

Reviewer 2: Unless referring to a specific "game", I think we should refer to arbitary Bevy programs as "apps". So this should read:
"When you're making an app in Bevy, you will have a single `App`, which stores and handles ..."

Let's make a trivial Hello World app.

```rust
use bevy::prelude::*;

fn main(){
  App::build()
  .add_system(hello.system())
  .run();
}

fn hello(){
    println!("Hello, Bevy!")
}
```

So, we first create our [`App`](https://docs.rs/bevy/latest/bevy/app/struct.App.html) by telling it to [`build`](https://docs.rs/bevy/latest/bevy/app/struct.App.html#method.build), performing initialization steps.
This produces an [`AppBuilder`](https://docs.rs/bevy/latest/bevy/app/struct.AppBuilder.html), which we can use to configure the `App`.

> We do so by add a simple system, which prints "Hello, Bevy!" when it is run.

Reviewer 3: ```suggestion
We do so by add our `hello()` function as a system through a **trait extension method**, which will print 'Hello, Bevy!' when the app is run.
```
Explaining what `.system()` does is pretty important.

Author: Definitely agree. It will likely be fully deprecated for most use cases (including here) though, so this is probably not the best place to describe how the magic works.

Finally once all of our work is complete, we call [`AppBuilder::run()`](https://docs.rs/bevy/latest/bevy/app/struct.AppBuilder.html#method.run) to actually make our app *do things*.

## What makes an App?

So, what sort of data does are `App` really store?
Turns out: it's fairly simple.
Looking at the docs for [`App`](https://docs.rs/bevy/latest/bevy/app/struct.App.html), we find three fields:

> ```rust

Reviewer 2: I would prefer to not immediately throw stuff like `Box<dyn Fn(App) + 'static, Global>` at people on the second page of the Bevy Book. We can introduce World/Schedule/runner concepts without this.

By not inlining the App struct, we also insulate against future changes (such as the infrastructure required to support deferred plugin init).

Reviewer 4: Yes, I also talked with Alice about this. We ended up deciding to simply link to the API docs instead, where the user can also check the engine source code if they want. By the way, `App` fields are private implementation details and therefore should definitely not be in the Book.

struct App {
  world: World
  runner: Box<dyn Fn(App) + 'static, Global>
  schedule: Schedule
}
```

The `World` stores all of our data, both entity-component data and resources.
The `runner` field contains a function that tells our app how to loop and execute.

> And the `Schedule` contains all of the systems that execute on our data, stored in a way that our scheduler can use to run them at the appropriate time.

Reviewer 4: This is better suited as API docs. Especially as it is stated later:

> Generally, you'll be operating at a higher-level of abstraction than this

We can transfer this knowledge to the API docs, which are currently scarce. There is also a benefit in doing this: pointing users to the API docs will encourage them to to it by their own.

> Generally, you'll be operating at a higher-level of abstraction than this: thinking about data in terms of specific resources or components, adding systems to the right place rather than manipulating the `Schedule` yourself, and sticking to the default runner.

Reviewer 4: This is better suited as API docs. Especially as it is stated later:

> Generally, you'll be operating at a higher-level of abstraction than this

We can transfer this knowledge to the API docs, which are currently scarce. There is also a benefit in doing this: pointing users to the API docs will encourage them to to it by their own.

Author: Sounds good to me.
